core/interface.go

```diff
@@ -12,28 +11,40 @@ import (
 // Advertisement type
 type Advertisement schema.Advertisement
 
+// AdvLink type
+type AdvLink schema.Link_Advertisement
+
+// Index type
+type Index schema.Index
+
+// IndexLink type
+type IndexLink schema.Link_Index
+
 // Interface for a reference provider
 type Interface interface {
 	// PublishLocal provides a new advertisement locally.
-	PublishLocal(ctx context.Context, ad Advertisement) error
+	PublishLocal(ctx context.Context, adv Advertisement) (cid.Cid, error)
 
 	// Publish advertisement to indexer pubsub channel
 	// Every advertisement published to the pubsub channel
 	// is also provided locally.
-	Publish(ctx context.Context, ad Advertisement) error
+	Publish(ctx context.Context, adv Advertisement) error
 
 	// PushAdv pushes a new advertisement to a specific indexer
-	PushAdv(ctx context.Context, indexer peer.ID, ad Advertisement) error
+	PushAdv(ctx context.Context, indexer peer.ID, adv Advertisement) error
 
 	// Push an update for a single entry.
 	// This can be used to perform updates for a small number of CIDs
 	// When a full advertisement is not worth it (web3.storage case).
 	// Indexer may only accept pushes from authenticated providers.
-	Push(ctx context.Context, indexer peer.ID, cid cid.Cid, val indexer.Value)
+	Push(ctx context.Context, indexer peer.ID, cid cid.Cid, metadata []byte)
+
+	// PutEvent notifies that new data has been added to the provider
+	PutEvent(ctx context.Context, cids []cid.Cid, metadata []byte) error
 
-	// Put new content in provider's local index. Putting new content
-	// in the index triggers the generation of a new advertisement.
-	Put(cids []cid.Cid, val indexer.Value) (bool, error)
+	// RemoveEvent sends an event to the reference provider to notify
+	// that new data has been removed from the provider
+	RemoveEvent(ctx context.Context, cids []cid.Cid, metadata []byte) error
```

Comment:
Naming seems inconsistent with other functions. in this interface
Without docs, I would have thought this functions removes an event.

Ditto re `PutEvent`

maybe something like:

```suggestion
	NotifyRemoved(ctx context.Context, cids []cid.Cid, metadata []byte) error
```

Also provider design doc need to be updated though its PR is not merged yet :)

Reply:
Totally agree. Thanks @masih. I like this naming. 

I am afraid that several things may change while we develop the reference provider, so I'll defer updating the design doc for when we are comfortable with the state of the interface.